[studio-ui][4575] Create a PathNavigatorTree Widget

aachavarria opened the pull request:
Creating Epics, Reducers, Actions, Container, UI, etc

aachavarria committed:
Initial pathNavigatorTree.ts
Creating Epics, Reducers, Actions, Container, UI, etc

aachavarria committed:
adding actions to fetch item children: pathNavigatorTreeFetchPathChildren

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTreeUI.tsx:
```diff
@@ -0,0 +1,137 @@
+import Header from '../PathNavigator/PathNavigatorHeader';
+import AccordionDetails from '@material-ui/core/AccordionDetails';
+import CircularProgress from '@material-ui/core/CircularProgress';
+import { Typography } from '@material-ui/core';
```
Fix import

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTreeUI.tsx:
```diff
@@ -0,0 +1,137 @@
+  nodes: TreeNode;
+}
+
+function RenderTreeNode({ nodes }: RenderTreeNodeProps) {
```
RenderTreeNode => TreeNode

aachavarria committed:
Adding itemDisplay to pathNavigatorTree.ts
Updating pathNavigatorTree.ts reducer;

aachavarria committed:
Typo

aachavarria committed:
adding item Menu to Path Navigator Tree

aachavarria committed:
removing path from expanded state if have not children

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,197 @@
+    path: null,
+    anchorEl: null,
+    loaderItems: null
+  });
```
Use `useState` instead

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,197 @@
+export default function PathNavigatorTree(props: PathNavigatorTreeProps) {
+  const { label, id = props.label.replace(/\s/g, ''), rootPath, excludes, limit, icon, container } = props;
+  const state = useSelection((state) => state.pathNavigatorTree)[id];
+  const itemsByPath = useSelection((state) => state.content.items.byPath);
```
Pull & `useItemsByPath()`

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,197 @@
+    });
+  };
+
+  const onCloseItemMenu = () => setItemMenu({ ...itemMenu, path: null, anchorEl: null });
```
no need to spread; you only need to set anchorEl to null?

aachavarria replied:
it is to keep the number of loaderItems when the menu is being hidden

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,197 @@
+        numOfLoaderItems={itemMenu.loaderItems}
+        anchorEl={itemMenu.anchorEl}
+        onClose={onCloseItemMenu}
+      />
```
Should we use the GlobalDialogManager item menu?

aachavarria replied:
we are using this approach on PathNavigatorTree, the issue using the one on GlobalDialogManager is that we need to send anchor position

rart replied:
Let's try/discuss something like this...
```typescript
import { getOffsetTop, getOffsetLeft } from '@material-ui/core/Popover'

onOpenMenu = (e) => {
  const anchorRect = e.target.getBoundingClientRect();
  return {
    top: anchorRect.top + getOffsetTop(anchorRect, 'top'),
    left: anchorRect.left + getOffsetLeft(anchorRect, 'left'),
  }
}

```

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTreeItem.tsx:
```diff
@@ -0,0 +1,143 @@
+        iconContainer: classes.iconContainer
+      }}
+    >
+      {Array.isArray(node.children)
```
Looks unnecessary. When would it ever not be an array?

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTreeUI.tsx:
```diff
@@ -0,0 +1,138 @@
+  isCollapsed: boolean;
+  expandedNodes: string[];
+  classes?: Partial<Record<'root' | 'body', string>>;
+}
```
Keep type declarations right after imports

rart commented on ui/app/src/state/reducers/content.ts:
```diff
@@ -148,6 +152,19 @@ const reducer = createReducer<ContentState>(initialState, {
+          ...state.items.byPath,
+          [item.path]: item
+        }
+      }
```
Pull & merge. 
items got removed and items.byPath got moved up as itemsByPath

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTreeItem.tsx:
```diff
@@ -0,0 +1,143 @@
+  onLabelClick(path: string): void;
+  onIconClick(path: string): void;
+  onOpenItemMenu(element: Element, path: string): void;
+}
```
Keep type declarations right after imports

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTreeItem.tsx:
```diff
@@ -0,0 +1,143 @@
+
+import TreeItem from '@material-ui/lab/TreeItem';
+import React from 'react';
+import { TreeNode } from '../Navigation/FolderBrowserTreeView';
```
This guy should be extracted out to models

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTreeItem.tsx:
```diff
@@ -0,0 +1,143 @@
+                <MoreVertRoundedIcon fontSize="small" />
+              </IconButton>
+            </Tooltip>
+          </section>
```
Pull & checkout the tweaks on PathNavigator to avoid tooltip jumps/layout-issues

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4575

aachavarria committed:
adding preview click to the items

aachavarria committed:
adding filter to pathNavigatorTree.ts;

aachavarria committed:
adding Collapse to pathNavigatorTree.ts;

aachavarria committed:
Fixes with nested Filter

aachavarria committed:
removing unnecessary validation

aachavarria committed:
adding color to the filter icon

aachavarria committed:
Filtering Fixes

aachavarria committed:
Updates of keywordByPathRef

aachavarria committed:
Changes on pathNavigatorTreeUpdate to split it on pathNavigatorTreeCollapsePath and pathNavigatorTreeExpandPath

aachavarria committed:
Adding Pagination

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,365 @@
+  const dispatch = useDispatch();
+  useMount(() => {
+    dispatch(
+      pathNavigatorTreeInit({
```
let's use the createAction's second arg to retrieve/dispatch with stored state synchronously

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,365 @@
+    return () => {
+      subscription.unsubscribe();
+    };
+  }, [state, id, dispatch]);
```
State not a dependency of this effect

aachavarria replied:
now it is not, but it will be a dependency

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,365 @@
+            item: itemsByPath[path]
+          })
+        );
+      }
```
```suggestion
      dispatch(
        previewItem({
          item: itemsByPath[path],
          newTab: event.ctrlKey || event.metaKey
        })
      );
```

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,365 @@
+
+  useEffect(() => {
+    const subscription = onSearch$.pipe(debounceTime(400)).subscribe(({ keyword, path }) => {
+      fetchingPathsRef.current.push(path);
```
Check the item isn't in before including.

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,365 @@
+    fetchingPathsRef.current.push(path);
+    // setData({ ...nodesByPathRef.current[rootPath] });
+    dispatch(
+      pathNavigatorTreeFetchNextPathChildren({
```
Next?

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,365 @@
+        icon={icon}
+        container={container}
+        isCollapsed={state?.collapsed}
+        data={data}
```
data => rootNode

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTreeUI.tsx:
```diff
@@ -0,0 +1,158 @@
+  title: string;
+  icon?: Partial<StateStylingProps>;
+  container?: Partial<StateStylingProps>;
+  data: TreeNode;
```
rootNode

rart commented on ui/app/src/state/actions/pathNavigatorTree.ts:
```diff
@@ -0,0 +1,78 @@
+export const pathNavigatorTreeFetchNextPathChildrenFailed = /*#__PURE__*/ createAction<{
+  id: string;
+  error: Omit<AjaxError, 'request' | 'xhr'>;
+}>('PATH_NAVIGATOR_TREE_FETCH_NEXT_PATH_CHILDREN_FAILED');
```
What's the `Next` about?
What's the difference with `pathNavigatorTreeFetchPathChildren`?

rart commented on ui/app/src/state/reducers/pathNavigatorTree.ts:
```diff
@@ -0,0 +1,142 @@
+        [id]: {
+          ...state[id],
+          expanded: children.length
+            ? [...state[id].expanded]
```
Why would you renew/spread expanded if no change is occurring?
```suggestion
            ? state[id].expanded
```

rart commented on ui/app/src/state/reducers/pathNavigatorTree.ts:
```diff
@@ -0,0 +1,142 @@
+          expanded: children.length
+            ? [...state[id].expanded]
+            : options?.keyword
+            ? [...state[id].expanded]
```
Why would you renew/spread expanded if no change is occurring?
```suggestion
            ? state[id].expanded
```

rart commented on ui/app/src/state/reducers/pathNavigatorTree.ts:
```diff
@@ -0,0 +1,142 @@
+            ? [...state[id].expanded]
+            : options?.keyword
+            ? [...state[id].expanded]
+            : [...state[id].expanded.filter((path) => path !== parentPath)],
```
Filter returns a new array already
```suggestion
            : state[id].expanded.filter((path) => path !== parentPath),
```

aachavarria committed:
Adding PathNavigatorTree to ToolsPanel.tsx

aachavarria committed:
PR Feedback

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4575

aachavarria committed:
PR Feedback

aachavarria committed:
Path navigator tree local storage restore

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4575

aachavarria committed:
PathNavigatorTree.tsx Restoring State from localStorage;

aachavarria committed:
adding Skeleton loader

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,378 @@
+  const fetchingPathsRef = useRef([]);
+  const onSearch$ = useSubject<{ keyword: string; path: string }>();
+  const uiConfig = useSelection<GlobalState['uiConfig']>((state) => state.uiConfig);
+  const { expanded, collapsed } = getStoredPathNavigatorTree(site, user.username, id) ?? {};
```
This changes every render. This is better suited to be inside the init effect.

aachavarria replied:
im going to add a memo, because expanded needs to be used on the render

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,378 @@
+  useEffect(() => {
+    const nextFetching = [];
+    fetchingPathsRef.current.forEach((path) => {
+      if (childrenByParentPath && childrenByParentPath[path]) {
```
```suggestion
      if (childrenByParentPath?.[path]) {
```

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,378 @@
+            name: itemsByPath[childPath].label,
+            children: nodesByPathRef.current[childPath]?.children
+              ? nodesByPathRef.current[childPath].children
+              : [{ id: 'loading' }]
```
```suggestion
            children: nodesByPathRef.current[childPath]?.children ?? [{ id: 'loading' }]
```

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,378 @@
+  });
+  const { formatMessage } = useIntl();
+  const nodesByPathRef = useRef({});
+  const keywordByPathRef = useRef({});
```
What’s the use of this ref? you already have the keywordByPath state const.
Also keyword is a string. Why initialised as object?

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -0,0 +1,378 @@
+
+  const onCloseWidgetMenu = () => setWidgetMenu({ ...widgetMenu, anchorEl: null });
+
+  const onSimpleMenuClick = (option: string) => {
```
Rename to state what menu this click handler is for

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTreeItem.tsx:
```diff
@@ -0,0 +1,351 @@
+
+const translations = defineMessages({
+  filter: {
+    id: 'filter.placeholder',
```
scope the id to this component

rart commented on ui/app/src/models/GetChildrenByPathsOptions.ts:
```diff
@@ -0,0 +1,21 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+export interface GetChildrenByPathsOptions {
```
There are really all same options as the single request api. I think we can dispose this interface and use the original

rart commented on ui/app/src/state/epics/pathNavigatorTree.ts:
```diff
@@ -0,0 +1,146 @@
+    ),
+  // endregion
+  // region pathNavigatorTreeLocalStore
+  (action$, state$) =>
```
Is this epic in use?

aachavarria replied:
yes it is the one saving the state to localStorage, I added to new actions to this epic pathNavigatorTreeCollapsePath and pathNavigatorTreeExpandPath

rart commented on ui/app/src/state/reducers/content.ts:
```diff
@@ -124,6 +130,24 @@ const reducer = createReducer<ContentState>(initialState, {
+  [pathNavigatorTreeRestoreComplete.type]: (state, { payload }) => {
+    let nextByPath = {};
+    Object.values(payload.data).forEach((children) => {
+      nextByPath = { ...nextByPath, ...updateItemByPath(state, { payload: { parent: null, children } }).itemsByPath };
```
`updateItemByPath` does a little too much to be efficient on a loop. split out that logic or simply be specific here.

no need to resp read nextByPath, write directly on it via `Object.assign(nextByPath, freshChildreLookup)`

rart commented on ui/app/src/state/reducers/pathNavigatorTree.ts:
```diff
@@ -0,0 +1,162 @@
+        ...state,
+        [id]: {
+          ...state[id],
+          expanded: state[id].expanded.includes(path) ? [...state[id].expanded] : [...state[id].expanded, path]
```
pointless spread on the ? path

rart commented on ui/app/src/state/reducers/pathNavigatorTree.ts:
```diff
@@ -0,0 +1,162 @@
+        [id]: {
+          ...state[id],
+          ...(children.length === 0 &&
+            !options?.keyword && { expanded: state[id].expanded.filter((path) => path !== parentPath) }),
```
```suggestion
          expanded: children.length === 0 && !options?.keyword ? state[id].expanded.filter((path) => path !== parentPath) }) : state[id].expanded,
```

What's this logic for though? Why collapse the path when there aren't children or keyword?

rart replied:
I still think the ternary in the suggestion is more efficient than a spread of a single prop.

rart commented on ui/app/src/state/reducers/pathNavigatorTree.ts:
```diff
@@ -0,0 +1,162 @@
+        [id]: {
+          ...state[id],
+          // expanded,
+          // collapsed,
```
?

rart commented on ui/app/src/utils/array.ts:
```diff
@@ -40,8 +40,8 @@ export function asArray<T = unknown>(value: T | T[]): T[] {
 }
 
-export function createPresenceTable(list: string[]): LookupTable<boolean> {
+export function createPresenceTable(list: string[], defaultValue: any = true): LookupTable<any> {
```
```suggestion
export function createPresenceTable(list: string[], valueGenerator: (listItem: string) => any): LookupTable<any> {
```

rart commented on ui/app/src/utils/array.ts:
```diff
@@ -40,8 +40,8 @@ export function asArray<T = unknown>(value: T | T[]): T[] {
+export function createPresenceTable(list: string[], defaultValue: any = true): LookupTable<any> {
   const table = {};
-  list.forEach((value) => (table[value] = true));
+  list.forEach((value) => (table[value] = defaultValue));
```
```suggestion
  list.forEach((value) => (table[value] = valueGenerator ? valueGenerator(value) : true));
```

aachavarria committed:
#4575;
PR Feedback;
Fixing Issues related with LocalStorage Restore;

aachavarria committed:
#4575;
Fixing Issues related with LocalStorage Restore;

aachavarria committed:
#4575;
Adding FolderCreated Propagation;

aachavarria committed:
#4575;
Adding keyword local storage support;

rart commented on ui/app/src/utils/array.ts:
```diff
@@ -40,8 +40,8 @@ export function asArray<T = unknown>(value: T | T[]): T[] {
 }
 
-export function createPresenceTable(list: string[]): LookupTable<boolean> {
+export function createPresenceTable(list: string[], valueGenerator: (listItem: string) => any): LookupTable<any> {
```
```suggestion
export function createPresenceTable(list: string[]): LookupTable<boolean>;
export function createPresenceTable<T extends string | number | boolean>(list: string[], value: T): LookupTable<T>;
export function createPresenceTable<T extends any>(
  list: string[],
  valueGenerator: (listItem: string) => T
): LookupTable<T>;
export function createPresenceTable<T extends any = boolean>(
  list: string[],
  valueOrGenerator: string | number | boolean | ((listItem: string) => T) = true
): LookupTable<T> {
```

rart commented on ui/app/src/utils/array.ts:
```diff
@@ -40,8 +40,8 @@ export function asArray<T = unknown>(value: T | T[]): T[] {
+export function createPresenceTable(list: string[], valueGenerator: (listItem: string) => any): LookupTable<any> {
   const table = {};
-  list.forEach((value) => (table[value] = true));
+  list.forEach((value) => (table[value] = valueGenerator ? valueGenerator(value) : true));
```
```suggestion
  const callback =
    typeof valueOrGenerator === 'function'
      ? (value) => (table[value] = valueOrGenerator(table[value]))
      : (value) => (table[value] = valueOrGenerator);
  list.forEach(callback);
```

rart commented on ui/app/src/state/reducers/pathNavigatorTree.ts:
```diff
@@ -0,0 +1,164 @@
+      const children = {};
+      const total = {};
+      Object.keys(data).forEach((path) => {
+        children[path] = [...data[path].map((item) => item.path)];
```
Map returns a new array.
```suggestion
        children[path] = data[path].map((item) => item.path);
```

rart commented on ui/app/src/state/reducers/pathNavigatorTree.ts:
```diff
@@ -0,0 +1,164 @@
+        ...state,
+        [id]: {
+          ...state[id],
+          // if the expanded node has not children and is not a result from a filtering it means is a leaf node
```
```suggestion
          // If the expanded node has no children and is not filtered, it's a leaf node and there's no point keeping it in `expanded`
```

aachavarria committed:
#4575;
PR Feedback

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4575

aachavarria committed:
#4575;
Adding itemCreated event propagation

aachavarria committed:
#4497;
Adding missing events Propagation;

aachavarria committed:
#4497;
Adding LevelDescriptor;

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTreeItem.tsx:
```diff
@@ -0,0 +1,351 @@
+                  styles={{ root: { flexGrow: 1, height: '26px' } }}
+                  item={itemsByPath[node.id]}
+                  showPublishingTarget={true}
+                  showWorkflowState={true}
```
These are optional, you may also remove them all together.
```suggestion
                  showPublishingTarget
                  showWorkflowState
```

aachavarria committed:
#4497;
PR Feedback

aachavarria committed:
#4497;
Overflow issue on pathNavigatorTree.ts;
Adding pathNavigatorTree to uiConfigDefaults.ts;
